### PR TITLE
syncer(dm): first commit about another handle-error

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #5877 

### What is changed and how it works?

introduce streamModifier, please review unit tests about its usage.

#### Tests 

 - Unit test

### Release note 

```release-note
`None`.
```

### CODE REVIEW

**Comment by lance6716** [comment]
note the target branch is a feature branch

**Comment by lance6716** [comment]
image

good coverage :smile:

**Comment by lance6716** [comment]
https://github.com/pingcap/dm/issues/1249

should add an UT about this

File: dm/syncer/binlogstream/stream_modifier.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+	m.nextOp = 0
+	m.nextEventInOp = 0
+
+	for i, op := range m.ops {
+		if op.pos.Compare(loc.Position) < 0 {
+			m.nextOp = i + 1
+		} else {
+			break
+		}
+	}
```
Comment (D3Hunter) [inline dm/syncer/binlogstream/stream_modifier.go:262]: quite common operation to find the first index >= some location, maybe abstract into one method

**Review by D3Hunter — COMMENTED** [review]
will review test later

File: dm/syncer/binlogstream/stream_modifier.go [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+			toPrint = m.ops[i:]
+		}
+	}
+
+	reqs := make([]*pb.HandleWorkerErrorRequest, 0, len(toPrint))
+	for _, op := range toPrint {
+		reqs = append(reqs, op.originReq)
+	}
+
+	b, err2 := json.Marshal(reqs)
```
Comment (buchuitoudegou) [inline dm/syncer/binlogstream/stream_modifier.go:211]: Why is the `reqs` serialized into a JSON string rather than returned directly?

Comment (lance6716) [reply dm/syncer/binlogstream/stream_modifier.go:211]: We can return reqs and let caller do it. Will change it in this PR later.

File: dm/syncer/binlogstream/stream_modifier.go (hunk 1 above)
Comment (lance6716) [reply dm/syncer/binlogstream/stream_modifier.go:250]: https://github.com/pingcap/tiflow/pull/5890/commits/5f8f8a2b3fa2b06a3a0339b7ce8c0c54f5be1e39

File: dm/syncer/binlogstream/stream_modifier.go (hunk 2 above)
Comment (lance6716) [reply dm/syncer/binlogstream/stream_modifier.go]: https://github.com/pingcap/tiflow/pull/5890/commits/69ac38fcf5c4596a458f7bc3f43389cba57d7ba0

File: dm/syncer/binlogstream/stream_modifier.go (hunk 1 above)
Comment (lance6716) [reply dm/syncer/binlogstream/stream_modifier.go]: https://github.com/pingcap/tiflow/pull/5890/commits/00c29358a3420624f8150a2312da70603ab45b51

File: dm/syncer/binlogstream/stream_modifier.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+		e.Dump(buf)
+		events = append(events, buf.String())
+	}
+	return fmt.Sprintf("uuid: %s, op: %s, events: %s, originReq: %v", o.uuid, o.op, strings.Join(events, "\n"), o.originReq)
+}
+
+// streamModifier is not thread-safe.
+type streamModifier struct {
+	ops    []*operator // sorted on operator.Position
+	nextOp int         // next operator whose location is waiting to be matched
```
Comment (D3Hunter) [inline dm/syncer/binlogstream/stream_modifier.go:72]: why we keep ops before `nextOp`?

File: dm/syncer/binlogstream/stream_modifier.go [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+		if err != nil {
+			m.logger.DPanic("invalid position, should be verified in caller",
+				zap.String("position", posStr))
+			return []*pb.HandleWorkerErrorRequest{}
+		}
+
+		i := 0
+		for i < len(m.ops) && m.ops[i].pos.Compare(pos) < 0 {
+			i++
+		}
```
Comment (D3Hunter) [inline dm/syncer/binlogstream/stream_modifier.go:186]: why not use sort.Search too?

we can use a named method to make this common part is more clear

File: dm/syncer/binlogstream/stream_modifier.go (hunk 3 above)
Comment (lance6716) [reply dm/syncer/binlogstream/stream_modifier.go:72]: the op may need to be consumed again by sharding DDL resync. For example

position 10: inject [tbl1 DDL1, tbl1 DDL2]
position 11: inject ...
position 20: tbl2 DDL1
position 30: tbl2 DDL2

the consuming order is 10, 11, 20 (now finish sharding DDL1, resync), 10, 11, 20, 30. Assuming table checkpoint will filter DDLs of finished sharding groups.

global checkpoint is used to `RemoveOutdated`

File: dm/syncer/binlogstream/stream_modifier.go (hunk 4 above)
Comment (lance6716) [reply dm/syncer/binlogstream/stream_modifier.go]: https://github.com/pingcap/tiflow/pull/5890/commits/e828a097103347c27ce6d39081279901690738d8

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: dm/syncer/binlogstream/stream_modifier.go [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+// reset will also reset nextEventInOp to a correct value.
+func (m *streamModifier) reset(loc binlog.Location) {
+	m.nextOp = 0
+	m.nextEventInOp = 0
+
+	m.nextOp = sort.Search(len(m.ops), func(i int) bool {
+		return loc.Position.Compare(m.ops[i].pos) <= 0
+	})
```
[the remark this answers is not in the document]
Comment (lance6716) [reply dm/syncer/binlogstream/stream_modifier.go]: https://github.com/pingcap/tiflow/pull/5890/commits/b5aac57e5642dcc31fff5f5bfccfc9962d8fa804

**Review by D3Hunter — APPROVED** [review]

**Comment by lance6716** [comment]
the failed test is because the feature branch does not include https://github.com/pingcap/tiflow/pull/5916

/merge

**Pull request merged**